syntax = "proto2";

package pingfs.proto;

message BlockProto {
    optional uint64 id = 1;
    optional BlockDataProto data = 2;
}

message BlockDataProto {
    oneof data {
        DirProto dir = 1;
        LinkProto link = 2;
        FileStartProto file_start = 3;
        FileContentsProto file_contents = 4;
    }
}

message DirProto {
    optional string dirname = 1;
    repeated uint64 children = 2;
}

message LinkProto {
    repeated uint64 children = 1;
}

message FileStartProto {
    optional string filename = 1;
    repeated uint64 children = 2;
}

message FileContentsProto {
    optional bytes data = 1;
}

enum ReadWriteExecuteProto {
    UNKNOWN = 1;
    NONE = 2;
    READ = 3;
    READ_WRITE = 4;
    READ_EXECUTE = 5;
    READ_WRITE_EXECUTE = 6;
    WRITE = 7;
    WRITE_EXECUTE = 8;
    EXECUTE = 9;
}
